spyder/plugins/ipythonconsole/widgets/tests/test_kernelconnect.py

```diff
@@ -0,0 +1,89 @@
+# -*- coding: utf-8 -*-
+# -----------------------------------------------------------------------------
+# Copyright © Spyder Project Contributors
+#
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+# -----------------------------------------------------------------------------
+
+"""Tests for the Existing Kernel Connection widget."""
+
+# Third party imports
+import pytest
+from qtpy.QtCore import Qt
+from qtpy.QtWidgets import QDialogButtonBox
+import time
+
+# Local imports
+from spyder.plugins.ipythonconsole.widgets.kernelconnect import (
+    KernelConnectionDialog
+)
+from spyder.config.main import CONF
+
+
+# =============================================================================
+# Fixtures
+# =============================================================================
+def setup_dialog(qtbot):
+    """Set up kernel connection dialog."""
+    widget = KernelConnectionDialog()
+    qtbot.addWidget(widget)
+
+    return widget
+
+
+# =============================================================================
+# Tests
+# =============================================================================
+def test_save(qtbot):
+    dlg = setup_dialog(qtbot)
+
+    cf_path = "test_cf_path"
+    un = "test_username"
+    hn = "test_hostname"
+    pn = 123
+    kf = "test_kf"
+    kfp = "test_kfp"
+
+    # fill out cf path
+    qtbot.keyClicks(dlg.cf, cf_path)
+    # check the remote kernel box
+    dlg.rm_group.setChecked(True)
+    # fill out hostname
+    qtbot.keyClicks(dlg.hn, hn)
+    # fill out port
+    dlg.pn.clear()
+    qtbot.keyClicks(dlg.pn, str(pn))
+    # fill out username
+    qtbot.keyClicks(dlg.un, un)
+
+    # select ssh keyfile radio
+    dlg.kf_radio.setChecked(True)
+    assert dlg.kf.isEnabled()
+
+    # fill out ssh keyfile
+    qtbot.keyClicks(dlg.kf, kf)
+
+    # fill out passphrase
+    dlg.kfp.clear()
+    qtbot.keyClicks(dlg.kfp, kfp)
+
+    # check save connection settings
+    dlg.save_layout.setChecked(True)
+
+    # save connection settings
+    dlg.save_connection_settings()
```

Comment:
So the problem here is that the normal code path calls the static method `get_connection_parameters()` which calls `exec_()` on the dialog and waits for user input. It seems like once I call `get_connection_parameters()` from the test code, the next line only executes after I manually dismiss the dialog.

Reply:
> Hmm but get_connection_parameters() is the method that calls save_connection_settings() if it's accepted.

You could just call ``self.accept_btns.accepted.connect(self.save_connection_settings)`` [here](self.accept_btns.accepted.connect(self.accept)) right before the connection to ``self.accept``, instead of doing it afterward outside the dialog, so clicking OK saves the dialog contents first. You'd also need to include the check for ``self.save_layout.ischecked()`` as the first line of save_connection_settings(self) if you went that route, ``return``ing if not.

Saving the connection settings ideally should be presumably done as soon as the OK button is clicked, to prevent something else from going wrong first and the settings not being saved, as well as since (IMO) it is a property of the dialog itself, and should automatically be done whenever the dialog is accepted and the box is ticked, rather than being something external and specific to a way of opening the dialog—but I'm not a Qt or even really OOP expert, so perhaps @ccordoba12 or @jitseniesen know better than I do if this is really a good idea or not.

Alternatively (or in addition), you could just make the dialog non-modal (as it should be anyway) by calling ``.show()`` instead of ``.exec_()`` in ``get_connection_settings()`` but @ccordoba12 might consider that out of scope for this PR.

> (I must admit I'm not a full time python developer :p)

Neither am I, if you couldn't tel already!